daffodil-io/src/main/scala/org/apache/daffodil/io/InputSource.scala

```diff
@@ -255,18 +292,27 @@ class BucketingInputSource(
    */
   private def fillBucketsToIndex(goalBucketIndex: Long, bytesNeededInBucket: Long): Boolean = {
     var lastBucketIndex = buckets.length - 1
-    var hasMoreData = true
+
     var needsMoreData = goalBucketIndex > lastBucketIndex || (goalBucketIndex == lastBucketIndex && bytesNeededInBucket > bytesFilledInLastBucket)
 
     while (needsMoreData && hasMoreData) {
       // Try to fill the rest of this bucket, regardless of how many bytes are
       // actually needed
       val emptyBytesInLastBucket = bucketSize - bytesFilledInLastBucket
 
+      Assert.invariant(emptyBytesInLastBucket >= 0)
```

Comment:
Are there any errors if we change this assert to > instead of >=? I don't think we should ever be calling read with a len of zero bytes.